File: packages/javascript/lib/format/duration.ts
import type { DurationOptions } from '../_types/global.js';
import { durationParts, findUnit } from './durationParts.js';

function label(value: number, name: string, useShortString: boolean, useSpace: boolean): string {
	const unit = findUnit(name);
	const space = useSpace ? ' ' : '';
	const suffix = useShortString ? (unit?.short ?? name) : `${name}${value === 1 ? '' : 's'}`;

	return `${value}${space}${suffix}`;
}

export function duration(milliseconds: number, options?: DurationOptions): string {
	const { useShortString = false, useSpace = true, separator = ' ' } = { ...options };

	return durationParts(milliseconds, options)
		.map((part) => label(part.value, part.unit, useShortString, useSpace))
		.join(separator);
}

File: packages/python/qsu/format/duration.py
from typing import Optional

from .durationParts import _findUnit, durationParts


def _numStr(value) -> str:
	if isinstance(value, float) and value.is_integer():
		return str(int(value))
	return str(value)


def _label(value, name: str, useShortString: bool, useSpace: bool) -> str:
	unit = _findUnit(name)
	space = ' ' if useSpace else ''

	if useShortString:
		suffix = unit['short'] if unit else name
	else:
		suffix = f"{name}{'' if value == 1 else 's'}"

	return f'{_numStr(value)}{space}{suffix}'


def duration(milliseconds: float, options: Optional[dict] = None) -> str:
	opts = {**(options or {})}
	useShortString = opts.get('useShortString', False)
	useSpace = opts.get('useSpace', True)
	separator = opts.get('separator', ' ')

	return separator.join(
		_label(part['value'], part['unit'], useShortString, useSpace)
		for part in durationParts(milliseconds, opts)
	)
